onnxruntime/core/providers/cpu/tensor/gather_elements.cc

```diff
@@ -88,153 +86,84 @@ static inline void increment_over_inner_dim(std::vector<int64_t>& current_dims,
-                         (GetNegativeIndexAdjustedValue<Tin>(indices_data, static_cast<Tin>(i) + indices_counter, axis, input_shape) *
-                          input_shape_pitches[axis]) +
-                         i];
-        };
-        conditional_batch_call(fn);
-        output_counter += inner_dim_size;
-      } else {
-        auto fn = [input_data, output_data, base_offset, input_shape_pitches, element_size,
-                   indices_data, indices_counter, axis, input_shape](ptrdiff_t i) {
-          memcpy(output_data + (i * element_size),
-                 input_data + (base_offset + (GetNegativeIndexAdjustedValue<Tin>(indices_data, static_cast<Tin>(i) + indices_counter, axis, input_shape) * input_shape_pitches[axis]) + i) * element_size,
-                 element_size);
-        };
-        conditional_batch_call(fn);
-        output_data += inner_dim_size * element_size;
-      }
-      indices_counter += static_cast<Tin>(inner_dim_size);
-      increment_over_inner_dim(process_dims, indices_shape);
-    }
-  }
-  // we special-case inner dim as we can weed-out some unnecessary computations in element offset calculations
-  else {
+  auto MainLoop = [&](auto* output, auto* input, auto LoopFn) {
     while (num_inner_dim-- != 0) {
-      base_offset = compute_base_offset(process_dims, input_shape_pitches, axis);
-
-      // process 1 chunk of 'inner dimension' length
-      if (is_string) {
-        auto fn = [input_data, output_data, base_offset,
-                   indices_data, indices_counter, axis, input_shape, output_counter](ptrdiff_t i) {
-          // for innermost axis, input_shape_pitches[axis] = 1 (so no need to multiply)
-          output_data[i + output_counter] =
-              input_data[base_offset +
-                         GetNegativeIndexAdjustedValue<Tin>(indices_data, static_cast<Tin>(i) + indices_counter, axis, input_shape)];
-        };
-        conditional_batch_call(fn);
-        output_counter += inner_dim_size;
-      } else {
-        // for innermost axis, input_shape_pitches[axis] = 1 (so no need to multiply)
-        auto fn = [input_data, output_data, base_offset, element_size,
-                   indices_data, indices_counter, axis, input_shape](ptrdiff_t i) {
-          memcpy(output_data + (i * element_size),
-                 input_data + (base_offset +
-                               GetNegativeIndexAdjustedValue<Tin>(indices_data, static_cast<Tin>(i) + indices_counter, axis, input_shape)) *
-                                  element_size,
-                 element_size);
-        };
-        conditional_batch_call(fn);
-        output_data += inner_dim_size * element_size;
-      }
-
-      indices_counter += static_cast<Tin>(inner_dim_size);
+      LoopFn(output, input + compute_base_offset(process_dims, input_shape_pitches, axis));
+      output += inner_dim_size;
+      indices_data += static_cast<Tin>(inner_dim_size);
       increment_over_inner_dim(process_dims, indices_shape);
```

Comment:
looks like we don't need to compute process_dims. all we need to know is the index of axis dimension. We can use an offset to track where we handled and increase it by:
offset += inner_dim_size;
input += inner_dim_size;
And index of axis dimension: index_axis = (offset / axis_pitch) % axis_size;

and output[i] = input[(GetIndex(i, indices_data, axis_size) - index_axis) * axis_pitch + i];

Reply:
Does that work even if the indices are a subset of the input tensor? I thought it was still required since the input/output pitches won't necessarily match due to this.